cpp/src/parquet/chunker_internal_test.cc

```diff
@@ -0,0 +1,908 @@
+};
+
+PageSizes GetColumnPageSizes(const std::shared_ptr<Buffer>& data, int column_index = 0) {
+  // Read the parquet data out of the buffer and get the sizes and lengths of the
+  // data pages in given column. We assert on the sizes and lengths of the pages
+  // to ensure that the chunking is done correctly.
+  PageSizes result;
+
+  auto buffer_reader = std::make_shared<BufferReader>(data);
+  auto parquet_reader = ParquetFileReader::Open(std::move(buffer_reader));
+
+  auto metadata = parquet_reader->metadata();
+  for (int rg = 0; rg < metadata->num_row_groups(); rg++) {
+    auto page_reader = parquet_reader->RowGroup(rg)->GetColumnPageReader(column_index);
+    while (auto page = page_reader->NextPage()) {
+      if (page->type() == PageType::DATA_PAGE || page->type() == PageType::DATA_PAGE_V2) {
+        auto data_page = static_cast<DataPage*>(page.get());
+        result.sizes.push_back(data_page->size());
+        result.lengths.push_back(data_page->num_values());
+      }
+    }
+  }
+
+  return result;
+}
+
+Result<PageSizes> WriteAndGetPageSizes(const std::shared_ptr<Table>& table,
+                                       uint64_t min_chunk_size, uint64_t max_chunk_size,
+                                       bool enable_dictionary = false,
+                                       int column_index = 0) {
+  // Write the table to a buffer and read it back to get the page sizes
+  ARROW_ASSIGN_OR_RAISE(
+      auto buffer,
+      WriteTableToBuffer(table, min_chunk_size, max_chunk_size, enable_dictionary));
+  ARROW_ASSIGN_OR_RAISE(auto readback, ReadTableFromBuffer(buffer));
+
+  RETURN_NOT_OK(readback->ValidateFull());
+  if (readback->schema()->Equals(*table->schema())) {
+    ARROW_RETURN_IF(!readback->Equals(*table),
+                    Status::Invalid("Readback table not equal to original"));
+  }
+  return GetColumnPageSizes(buffer, column_index);
+}
+
+void AssertAllBetween(const std::vector<uint64_t>& values, uint64_t min, uint64_t max,
+                      bool expect_dictionary_fallback = false) {
+  // expect the last chunk since it is not guaranteed to be within the range
+  if (expect_dictionary_fallback) {
+    // if dictionary encoding is enabled, the writer can fallback to plain
+    // encoding splitting within a content defined chunk, so we can't
+    // guarantee that all chunks are within the range in this case, but we
+    // know that there can be at most 2 pages smaller than the min_chunk_size
+    size_t smaller_count = 0;
+    for (size_t i = 0; i < values.size() - 1; i++) {
+      if (values[i] < min) {
+        smaller_count++;
+      } else {
+        ASSERT_LE(values[i], max);
+      }
+    }
+    ASSERT_LE(smaller_count, 2);
+  } else {
+    for (size_t i = 0; i < values.size() - 1; i++) {
+      ASSERT_GE(values[i], min);
+      ASSERT_LE(values[i], max);
+    }
+  }
+  ASSERT_LE(values.back(), max);
+}
+
+std::vector<std::pair<std::vector<uint64_t>, std::vector<uint64_t>>> FindDifferences(
```

Comment:
The `std::vector<std::pair<std::vector<uint64_t>, std::vector<uint64_t>>>` is gonna be a bit cryptic, can you instead define struct types with meaningful names?

Reply:
Updated, added ChunkList, ChunkDiff types for better readability.